### PR TITLE
Add cohort factor

### PR DESCRIPTION
**Pull request by chrispratt-tonyyeb**
I've used the role factor as a template to create a factor based on cohorts.

### CODE REVIEW

**Comment by chrispratt-tonyyeb** [comment]
#358 Will resolve this

File: factor/cohort/lang/en/factor_cohort.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * Language strings.
+ *
+ * @package     factor_cohort
+ * @author      Chris Pratt <tonyyeb@gmail.com>
+ * @copyright   Chris Pratt
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+$string['pluginname'] = 'Cohort Factor';
+$string['privacy:metadata'] = 'The user capability factor plugin does not store any personal data';
```
Comment (Peterburnett) [inline factor/cohort/lang/en/factor_cohort.php:27]: String references wrong factor here

File: factor/cohort/settings.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * @author      Chris Pratt <tonyyeb@gmail.com>
+ * @copyright   Chris Pratt
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require_once(__DIR__ . '/../../../../../cohort/lib.php');
+
+defined('MOODLE_INTERNAL') || die();
+
+$settings->add(new admin_setting_configcheckbox('factor_cohort/enabled',
```
Comment (Peterburnett) [inline factor/cohort/settings.php:30]: Please implement the new callback step added in #349

File: factor/cohort/settings.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+$cohorts = cohort_get_all_cohorts();
+
+foreach($cohorts['cohorts'] as $cohort) {
+   $choices[$cohort->id] = $cohort->name;
+}
+
+$settings->add(new admin_setting_configmultiselect('factor_cohort/cohorts',
+    new lang_string('settings:cohort', 'factor_cohort'),
+    new lang_string('settings:cohort_help', 'factor_cohort'), NULL, $choices));
```
Comment (Peterburnett) [inline factor/cohort/settings.php:46]: Please use lowercase null, as per moodle coding standards

Comment (Peterburnett) [reply factor/cohort/settings.php:46]: https://docs.moodle.org/dev/Coding_style#Booleans_and_the_null_value

File: factor/cohort/classes/factor.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        global $USER;
+        $cohortstring = get_config('factor_cohort', 'cohorts');
+
+        // Nothing selected, everyone passes.
+        if (empty($cohortstring)) {
+            return \tool_mfa\plugininfo\factor::STATE_PASS;
+        }
+
+        $selected = explode(',', $cohortstring);
+
```
Comment (Peterburnett) [inline factor/cohort/classes/factor.php:87]: Can we compress these unnecessary whitespaces a bit?

File: factor/cohort/version.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * @package     factor_cohort
+ * @subpackage  tool_mfa
+ * @author      Chris Pratt <tonyyeb@gmail.com>
+ * @copyright   Chris Pratt
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+$plugin->version      = 2020072100;      // The current plugin version (Date: YYYYMMDDXX).
```
Comment (Peterburnett) [inline factor/cohort/version.php:29]: Please update the date here and the dependencies to the current MFA version number

**Review by Peterburnett — CHANGES_REQUESTED** [review]
Hi there, thanks for the contribution. This seems like a great factor and a useful addition to the plugin. Just a couple of comments I noticed blocking this one from landing.

**Comment by chrispratt-tonyyeb** [comment]
Hi @Peterburnett, thanks so much for your comments. As you'll see this was adapted from your role based factor and is my first fork/pull on github. I think I've covered everything you mentioned, have requested a re-review. Thanks again!

**Review by Peterburnett — APPROVED** [review]

**Comment by chrispratt-tonyyeb** [comment]
Hi @Peterburnett thanks for checking over the code again. I'm not sure how to resolve all those errors? Is that something I am responsible for? Thanks

File: factor/cohort/settings.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    \tool_mfa\manager::do_factor_action('cohort', get_config('factor_cohort', 'enabled') ? 'enable' : 'disable');
+});
+$settings->add($enabled);
+
+$settings->add(new admin_setting_configtext('factor_cohort/weight',
+    new lang_string('settings:weight', 'tool_mfa'),
+    new lang_string('settings:weight_help', 'tool_mfa'), 100, PARAM_INT));
+
+$cohorts = cohort_get_all_cohorts();
+
```
Comment (Peterburnett) [inline factor/cohort/settings.php:43]: CI is currently failing, as when there are no cohorts, choices is never initialised. Please initialise the $choices variable to an empty array before.

**Review by Peterburnett — CHANGES_REQUESTED** [review]
Just a small CI catch before merge

File: factor/cohort/settings.php (hunk 6 above)
Comment (chrispratt-tonyyeb) [reply factor/cohort/settings.php:44]: Ah yes, that's a good idea. Have added that. Can I ask where in github you could see that, please?

Comment (Peterburnett) [reply factor/cohort/settings.php:44]: image
In the actions run, if you expand out the moodle installation section

**Review by Peterburnett — CHANGES_REQUESTED** [review]
We are so close, just a small indentation error to make CI happy

image

File: factor/cohort/settings.php (hunk 6 above)
Comment (chrispratt-tonyyeb) [reply factor/cohort/settings.php:44]: Thank you for that, very helpful!

**Comment by chrispratt-tonyyeb** [comment]
All further changes completed

**Comment by chrispratt-tonyyeb** [comment]
> We are so close, just a small indentation error to make CI happy
> 
> image

These changes have been made, thanks

**Comment by Peterburnett** [comment]
Hi all,

I grabbed this branch and sorted the last CI issues, and merged them in #362. Thanks for your contribution @chrispratt-tonyyeb ! Closing this PR now

**Pull request closed without merging**